File: src/main/java/fr/deroffal/aoc/Day01.java
package fr.deroffal.aoc;

import java.util.stream.IntStream;

import fr.deroffal.aoc.util.Utils;

public class Day01 {

    public static void main(String[] args) {
        final String input = Utils.readLine("day01.txt");

        System.out.println("Partie 1 : " + calculerScore(input, 1));
        System.out.println("Partie 2 : " + calculerScore(input, input.length() / 2));
    }

    static int calculerScore(final String input, final int offset) {
        final int inputLenght = input.length();

        return IntStream.range(0, inputLenght).map(i -> {
            final char charAt = input.charAt(i);
            return charAt == input.charAt((i + offset) % inputLenght) ? Character.getNumericValue(charAt) : 0;
        }).sum();
    }
}

File: kotlin/src/main/kotlin/fr/deroffal/aoc/kotlin/Day01.kt
package fr.deroffal.aoc.kotlin

import fr.deroffal.aoc.kotlin.tools.getNumericValue

/**
 * @see <a href="https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/zip.html">Zip</a>
 */
fun main(args: Array<String>) {
    val input = fr.deroffal.aoc.kotlin.tools.readLine("day01.txt")
    val day01 = Day01()
    println("Part 1 : ${day01.solvePart1(input)}")
    println("Part 2 : ${day01.solvePart2(input)}")
}

class Day01 {

    internal fun solvePart1(input: String): Int = (input + input[0]).zipWithNext().filter { it.first == it.second }.map { it.first.getNumericValue() }.sum()

    internal fun solvePart2(input: String): Int {
        val half = input.length / 2
        return (input.substring(0, half) zip input.substring(half)).filter { it.first == it.second }.map { 2 * it.first.getNumericValue() }.sum()
    }

}
